### DISCUSSION TITLE
FormatException: Unknown entity type CastList<TableRow, Employee>

### DISCUSSION THREAD

**Discussion by tran-huy-phuc** [serverpod/serverpod#102] [Q&A] [answered]
I got above exception when implementing the .find( ) query. (I'm not sure if I missed any step)

1. Create employee.yaml 
```
class: Employee
table: employee
fields:
  name: String
  dob: DateTime
  role: String
  salary: double
```
2. Run `serverpod generate` to generate code
3. Execute the postgresql to create employee table
4. I already can create new employee(s). 
5. Implement the api to find employees:
```
class EmployeeEndpoint extends Endpoint {
  Future<Employee?> create(Session session, Employee employee) async {
    try {
      await Employee.insert(session, employee);
      return employee;
    } catch (e) {
      print(e);
      return null;
    }
  }

  Future<List<Employee>> fetchEmployees(Session session,
      {Expression Function(EmployeeTable)? where,
      int? limit,
      int? offset}) async {
    final result = await Employee.find(session,
        where: where, limit: limit, offset: offset);
    print(result);
    return result;
  }
}
```
6. Call the `fetchEmployees()` method on Flutter side, observe the json (from the print method above):
```
[{"class":"Employee","data":{"id":1,"name":"Phuc","dob":"2022-03-06T06:43:59.903927Z","role":"Engineer","salary":1000.0}}, {"class":"Employee","data":{"id":2,"name":"Phuc","dob":"2022-03-06T06:44:14.464248Z","role":"Engineer","salary":1000.0}}, {"class":"Employee","data":{"id":3,"name":"Phuc","dob":"2022-03-06T07:01:20.663493Z","role":"Engineer","salary":1000.0}}]
```
-> That means the code pulled the data correctly from database.

7. But at the last step it through the exception:
```
FormatException: Unknown entity type CastList<TableRow, Employee>
#0      SerializationManager.serializeEntity (package:serverpod_serialization/src/serialization.dart:87:7)
#1      Server._handleRequest (package:serverpod/src/server/server.dart:266:34)
<asynchronous suspension>
```

8. I did the debug into serializeEntity( ) method:
```
String? serializeEntity(dynamic entity) {
    if (entity == null) {
      return null;
    } else if (entity is String) {
      return jsonEncode(entity);
    } else if (entity is DateTime) {
      return entity.toIso8601String();
    } else if (entity is ByteData) {
      return entity.base64encodedString();
    } else if (entity is int ||
        entity is bool ||
        entity is double ||
        entity is SerializableEntity) {
      return '$entity';
    } else {
      throw FormatException('Unknown entity type ${entity.runtimeType}');
    }
  }
```
It went to else because the List (Employee) is not a SerializableEntity type.

**Comment by vlidholt** [comment, marked as answer]
Hi!

Only certain types can be passed as parameter to or returned from a method on an `Endpoint`. The docs describes it here: https://github.com/serverpod/serverpod#passing-parameters

There are some limitations to how endpoint methods can be implemented. Currently, named parameters are not yet supported. Parameters and return types can be of type `bool`, `int`, `double`, `String`, `DateTime`, or generated serializable objects (see next section). A typed `Future` should always be returned. Null safety is supported. When passing a `DateTime` it is always converted to UTC.

Returning a `List` is currently not supported, but will be in future versions. Your option now is to create a serializable class with a list as a field and return that from your endpoint method.

**Reply by tran-huy-phuc** [reply]
Thanks @vlidholt for your answer.

I just try serverpod experimentally now, so it's fine for me to wait until List is supported. 
But yeah, I understand the idea of creating a serializable class.

Thanks again for the great framework!

**Reply by anandsubbu007** [reply]
any update around this?

how to use find parameter in endpoints?

it is throwing FormatException: Unknown entity type CastList<TableRow, ClassModel>
